content/advent-2017/lexmachine-advent.md

```diff
@@ -0,0 +1,644 @@
+the second is the *lexing action* function. The action is called on pattern
+discovery. This action typically takes a low-level
+[\*machines.Match](https://godoc.org/github.com/timtadh/lexmachine/machines#Match)
+object and turns it into a token representation of your (the user's) choice.
+
+For users who do not have strict requirements for how tokens are represented
+the [\*lexmachine.Token](https://godoc.org/github.com/timtadh/lexmachine#Token)
+object provides a useful implementation. [\*lexmachine.Scanner](https://godoc.org/github.com/timtadh/lexmachine#Scanner) has a utility
+method for constructing the tokens which helps us write a simple `getToken`
+action:
+
+```go
+func getToken(tokenType int) lexmachine.Action {
+	return func(s *lexmachine.Scanner, m *machines.Match) (interface{}, error) {
+		return s.Token(tokenType, string(m.Bytes), m), nil
+	}
+}
+```
+
+### Adding all the patterns
+
+```go
+func newLexer() *lexmachine.Lexer {
+	lexer := lexmachine.NewLexer()
+	lexer.Add([]byte("@"), getToken(tokmap["AT"]))
+	lexer.Add([]byte(`\+`), getToken(tokmap["PLUS"]))
+	lexer.Add([]byte(`\*`), getToken(tokmap["STAR"]))
+	lexer.Add([]byte("-"), getToken(tokmap["DASH"]))
+	lexer.Add([]byte("/"), getToken(tokmap["SLASH"]))
+	lexer.Add([]byte("\\"), getToken(tokmap["BACKSLASH"]))
+	lexer.Add([]byte(`\^`), getToken(tokmap["CARROT"]))
+	lexer.Add([]byte("`"), getToken(tokmap["BACKTICK"]))
+	lexer.Add([]byte(","), getToken(tokmap["COMMA"]))
+	lexer.Add([]byte(`\(`), getToken(tokmap["LPAREN"]))
+	lexer.Add([]byte(`\)`), getToken(tokmap["RPAREN"]))
+	lexer.Add([]byte("bus"), getToken(tokmap["BUS"]))
+	lexer.Add([]byte("chip"), getToken(tokmap["CHIP"]))
+	lexer.Add([]byte("label"), getToken(tokmap["LABEL"]))
+	lexer.Add([]byte("compute"), getToken(tokmap["COMPUTE"]))
+	lexer.Add([]byte("ignore"), getToken(tokmap["IGNORE"]))
+	lexer.Add([]byte("set"), getToken(tokmap["SET"]))
+	lexer.Add([]byte(`[0-9]*\.?[0-9]+`), getToken(tokmap["NUMBER"]))
+	lexer.Add([]byte(`[a-zA-Z_][a-zA-Z0-9_]*`), getToken(tokmap["NAME"]))
+	lexer.Add([]byte(`"((\\.)|([^"\\]))*"`), func(s *lexmachine.Scanner, m *machines.Match) (interface{}, error) {
+		return s.Token(tokmap["NAME"], string(m.Bytes[1:len(m.Bytes)-1]), m), nil
+	})
+	lexer.Add([]byte(`#[^\n]*`), getToken(tokmap["COMMENT"]))
+	lexer.Add([]byte(`( |\t|\f)+`), getToken(tokmap["SPACE"]))
+	lexer.Add([]byte(`\\\n`), getToken(tokmap["SPACE"]))
+	lexer.Add([]byte(`\n|\r|\n\r`), getToken(tokmap["NEWLINE"]))
+	err = lexer.Compile()
+	if err != nil {
+		panic(err)
+	}
+	return lexer
+}
+```
+
+### Skipping a pattern
+
+Sometimes it is advantageous to not emit tokens for certain patterns and to
+instead skip them. Commonly this occurs for whitespace and comments.  To skip a
+pattern simply have the action `return nil, nil`:
+
+```go
+func skip(scan *Scanner, match *machines.Match) (interface{}, error) {
+	return nil, nil
+}
+```
+
+```go
+lexer.Add([]byte("\s+"), skip)     // skip whitespace
```

Comment:
I think this is a compile error. You need ` quotes.

Reply:
good catch.